test/gtest/rope.hpp

```diff
@@ -0,0 +1,318 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2024 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#include "../driver/tensor_driver.hpp"
+#include "get_handle.hpp"
+#include "random.hpp"
+#include "tensor_holder.hpp"
+#include "verify.hpp"
+#include <gtest/gtest.h>
+#include <miopen/rope.hpp>
+#include <miopen/miopen.h>
+
+template <class T>
+void cpu_rope_forward(tensor<T> x, tensor<T> cos, tensor<T> sin, tensor<T>& ref_y)
+{
+    auto x_dims   = x.desc.GetLengths();
+    auto cos_dims = cos.desc.GetLengths();
+    auto input_numel =
+        std::accumulate(x_dims.begin(), x_dims.end(), 1L, std::multiplies<int64_t>());
+    auto rotary_numel =
+        std::accumulate(cos_dims.begin(), cos_dims.end(), 1L, std::multiplies<int64_t>());
+
+    par_ford(input_numel)([&](size_t o) {
+        size_t freq_idx = o % rotary_numel;
+
+        T input             = static_cast<T>(x[o]);
+        T input_rotate_half = (o % 2 == 0) ? static_cast<T>(-x[o + 1]) : static_cast<T>(x[o - 1]);
+
+        T cos_val = static_cast<T>(cos[freq_idx]);
+        T sin_val = static_cast<T>(sin[freq_idx]);
+
+        T val = (input * cos_val) + (input_rotate_half * sin_val);
+
+        ref_y[o] = val;
+    });
+}
```

Comment:
[not related to this particular PR]
@atamazov @JehandadKhan is it by-design that we have 2 CPU implementations - one for the driver one for the test?

Reply:
@CAHEK7 No.